official/resnet/resnet_run_loop.py

```diff
@@ -86,21 +98,23 @@ def process_record_dataset(dataset, is_training, batch_size, shuffle_buffer,
   # last batch of an epoch-- do not raise an error when we try to split them
   # over the GPUs. This will likely be handled by Estimator during replication
   # in the future, but for now, we just drop the leftovers here.
-  if multi_gpu:
+  if use_distribution_strategy:
     total_examples = num_epochs * examples_per_epoch
-    dataset = dataset.take(batch_size * (total_examples // batch_size))
+    dataset = dataset.take(
+        per_device_batch_size * (total_examples // per_device_batch_size))
 
   # Parse the raw records into images and labels
-  dataset = dataset.map(lambda value: parse_record_fn(value, is_training),
-                        num_parallel_calls=num_parallel_calls)
-
-  dataset = dataset.batch(batch_size)
+  dataset = dataset.apply(
+      tf.contrib.data.map_and_batch(
+          lambda value: parse_record_fn(value, is_training),
+          batch_size=per_device_batch_size,
+          num_parallel_batches=1))
```

Comment:
Why is num_parallel_batches set to 1 (wouldn't more improve performance)?

Reply:
It still parallelizes the lambda call. So we probably only need >1 if there are stragglers, n_cores > batch_size.

Reply:
if you want to make this more generic, the ideal thing is num_parallel_batches = num_cores / batch_size
Since usually batch_size > num_cores, i used 1. But yeah you could do that division and take the cieling or something?

Reply:
Interestingly, when I had tried this with replicate_model_fn alone, it actually slowed performance. Same for tf_cnn. See https://github.com/tensorflow/benchmarks/issues/137 , and the abandoned branch in which I attempted- https://github.com/tensorflow/models/compare/feat/contrib-data . This is JFYI assuming that DistStrat gets better performance, but not a bad idea to take in num_cpus in any case, as Transformer uses that as well, will be useful elsewhere.

Reply:
Yes, that's definitely interesting. We saw a performance bump both in OneDeviceStrategy and MirroredStrategy with this approach.. agreed that it might still be worthwhile to take in the num cores and use that to get the num_parallel_batches

Reply:
I just tested this on a trivial imagenet resnet (so any limit is the input pipeline), and found no difference with num_parallel_batches. (This is 32 cores, 4xP100, batch_size=512.) I'm hard pressed to think of a case where num_cores < batch_size makes sense, so I'm inclined to leave it at 1 to keep the code simple.

Reply:
That sounds good to me if you add a comment describing the other approach. CC @mrry in case you have thoughts.

Reply:
Setting it to 1 seems like a fine way to go. Any larger would probably lead to congestion on the threadpool queues. The ideal number for a batch size of 512 is probably between 0 and 1, and there's an outstanding bug to support more precise control of the parallelism here. /cc @jsimsa